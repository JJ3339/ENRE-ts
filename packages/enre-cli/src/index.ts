import usingCore, {preferences} from '@enre/core';
import cli from './cli';
import lsifDumper from './dumper/lsif';

cli.parse(process.argv);
const opts = cli.opts();

// Override default preferences according to command-line input
preferences.set('performance.multi-thread-enabled', opts.multiThread);
preferences.set('logging.verbose', opts.verbose);
preferences.set('performance.number-of-processors', parseInt(process.env.NUMBER_OF_PROCESSORS || '1'));
preferences.set('info.base-path', process.cwd());

// Trigger core functionality to do the analysis
await usingCore(opts.input, opts.exclude);

lsifDumper(opts);
