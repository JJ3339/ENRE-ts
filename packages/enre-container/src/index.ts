/**
 * ENRE entity container
 */
export {default as eGraph, ENREEntityPredicates} from './container/eContainer';

/**
 * ENRE entities
 */
export type {ENREEntityBase} from './entity/Base';

export type {ENREEntityClass} from './entity/Class';
export {recordEntityClass} from './entity/Class';

export type {ENREEntityEnum} from './entity/Enum';
export {recordEntityEnum} from './entity/Enum';

export type {ENREEntityEnumMember} from './entity/EnumMember';
export {recordEntityEnumMember} from './entity/EnumMember';

export type {ENREEntityField} from './entity/Field';
export {recordEntityField} from './entity/Field';

export type {ENREEntityFile} from './entity/File';
export {recordEntityFile} from './entity/File';

export type {ENREEntityFunction} from './entity/Function';
export {recordEntityFunction} from './entity/Function';

export type {ENREEntityInterface} from './entity/Interface';
export {recordEntityInterface} from './entity/Interface';

export type {ENREEntityMethod} from './entity/Method';
export {recordEntityMethod} from './entity/Method';

export type {ENREEntityParameter} from './entity/Parameter';
export {recordEntityParameter} from './entity/Parameter';

export type {ENREEntityVariable, ENREEntityVariableKind} from './entity/Variable';
export {recordEntityVariable} from './entity/Variable';

/**
 * ENRE entity collections
 */
export type {
  ENREEntityCollectionInFile, ENREEntityCollectionAll, ENREEntityCollectionScoping, ENREEntityTypes
} from './entity/collections';

/**
 * ENRE relation container
 */
export {default as rGraph, ENRERelationPredicates} from './container/rContainer';

/**
 * ENRE relations
 */
export type {ENRERelationExtend} from './relation/Extend';
export {recordRelationExtend} from './relation/Extend';

export type {ENRERelationImplement} from './relation/Implement';
export {recordRelationImplement} from './relation/Implement';

/**
 * ENRE relation collections
 */
export type {
  ENRERelationCollectionAll, ENRERelationTypes
} from './relation/collections';
