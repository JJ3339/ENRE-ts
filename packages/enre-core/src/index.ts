import {eGraph, ENREEntityFile, recordEntityFile} from '@enre/container';
import {warn} from '@enre/logging';
import path from 'path';
import {analyse} from './analyser';
import linker from './analyser/linker';
import {getFileList} from './utils/fileFinder';
import preferences from './utils/preferences';
import {buildENREName, ENRENameFile} from '@enre/naming';

/**
 * This injects a function 'formatENRE' to the built-in String's prototype for convenience.
 */
import './utils/add-string-format-to-string-proto';

export default async (
  iPath: string,
  exclude: Array<string> | undefined = undefined
) => {
  // Init global storages


  const fl = await getFileList(iPath, exclude);

  // Suggest multi thread functionality if not enabled
  if (!preferences.get('performance.multi-thread-enabled') && (preferences.get('performance.number-of-processors') > 1)) {
    // info('Multi cores processor detected, running with tag \'-m\' can improve performance significantly');
  }

  if (preferences.get('performance.multi-thread-enabled')) {
    warn('Multi-threading is currently disabled, you have set preference wrong');
    // const sPool = new StaticPool({
    //   size: Math.min(fl.length, preferences.get('performance.number-of-processors')),
    //   task: './src/parser-mt.js'
    // });
    //
    // fl.forEach(record => {
    //   (async () => {
    //     const ast = await sPool.exec(record);
    //   })();
    // });
  } else {
    /**
     * PRE PASS: Create file entities for every entry.
     * TODO: Remove this when dedicated package & file extractor is done.
     */
    for (const f of fl) {
      const fParsed = path.parse(f);
      recordEntityFile(
        // TODO: fix hard code
        buildENREName<ENRENameFile>({base: fParsed.name, ext: fParsed.ext.substring(1) as 'js'}),
        [path.dirname(f)],
        // TODO: sourceType detect
        'module',
        // TODO: Migrate lang to ts enum
        path.extname(f).includes('ts') ? 'ts' : 'js');
    }

    /**
     * FIRST PASS: Extract entities and immediate relations, build entity graph.
     */
    for (const f of eGraph.where({type: 'file'})) {
      await analyse(f as ENREEntityFile);
    }

    /**
     * SECOND PASS: Work on pseudo relation container to link string into
     * correlated entity object.
     */
    linker();

    /**
     * THIRD PASS: Based on full entity-relation graph,
     * extracting relations that depends on full data
     * (like 'override').
     */
    // TODO: advancedLinker();

    // if (!environment.test) {
    //   const groupingEntities = eGraph.all
    //     .reduce((
    //       prev: Partial<Record<ENREEntityTypes, Array<ENREEntityCollectionAll>>>,
    //       curr
    //     ) => {
    //       prev[curr.type]?.push(curr) || (prev[curr.type] = [curr]);
    //       return prev;
    //     }, {});
    //   Object.keys(groupingEntities)
    //     .forEach(k => console.log(`Entity ${k}: ${groupingEntities[k as ENREEntityCollectionAll['type']]!.length}`));
    //
    //   const groupingRelations = rGraph.all
    //     .reduce((
    //       prev: Partial<Record<ENRERelationTypes, Array<ENRERelationCollectionAll>>>,
    //       curr
    //     ) => {
    //       prev[curr.type]?.push(curr) || (prev[curr.type] = [curr]);
    //       return prev;
    //     }, {});
    //   Object.keys(groupingRelations)
    //     .forEach(k => console.log(`Relation ${k}: ${groupingRelations[k as ENRERelationCollectionAll['type']]!.length}`));
    // }
  }
};

export {default as preferences} from './utils/preferences';
export {cleanAnalysis} from './analyser';
