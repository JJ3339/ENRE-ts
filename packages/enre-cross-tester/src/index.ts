const {promisify} = await import('util');
const exec = promisify((await import('child_process')).exec);

export const available = async () => {
  try {
    const {stdout} = await exec('upython -V');
    return /Python 3\.\d+\.\d+/.test(stdout);
  } catch (e) {
    return false;
  }
};

export default async (dbpath) => {
  try {
    if (!await available()) {
      console.log('upython is not available, please check system\' environment path');
      return undefined;
    }
    const {stdout} = await exec(`upython cross/do.py -p ${dbpath} null`);
    return rebuildObjectGraph(JSON.parse(stdout));
  } catch (e) {
    console.log(e.message);
  }
};

/**
 * Convert numbered ID into reference to that entity object
 */
const rebuildObjectGraph = (raw) => {
  switch (raw['script_ver']) {
  case '1.0.0':
    const entityLookup = new Map();
    for (const ent of raw['entities']) {
      if (ent['type'] === 'File') {
        ent['entities'] = [];
        ent['relations'] = [];
        entityLookup.set(ent['id'], ent);
      } else {
        ent['relations'] = [];
        entityLookup.set(ent['id'], ent);
        entityLookup.get(ent['belongs_to']).entities.push(ent);
        const testAnonymity = /\(unnamed_(class|function)_\d+\)/.test(ent['name']);
        if (testAnonymity !== null) {
          ent['name'] = {isAnonymous: true, as: testAnonymity[1]};
        }
      }
    }
    for (const rel of raw['relations']) {
      rel['to'] = entityLookup.get(rel['to']);
      entityLookup.get(rel['from']).relations.push(rel);
    }
    break;
  default:
    console.log(`Unhandled script version ${raw['script_ver']}`);
    return undefined;
  }

  return raw;
};
