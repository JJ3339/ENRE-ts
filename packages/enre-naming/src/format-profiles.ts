/**
 * Format profile defines how ENRE output entity's name
 * if it is not a usual identifier.
 */

export enum FormatProfile {
  /**
   * **Default format**
   *
   * This will output anonymous entity and invalid identifier
   * as an XML string used in docs.
   */
  default,
  /**
   * **Understand format**
   *
   * This will output invalid identifier using SciTools Understand's format.
   */
  understand,
  /**
   * **Node format**
   *
   * This will output invalid identifier using node's format.
   */
  node,
}

export let currentProfile = FormatProfile.default;

export function usingNewFormatProfile(newFormat: FormatProfile) {
  currentProfile = newFormat;
}
