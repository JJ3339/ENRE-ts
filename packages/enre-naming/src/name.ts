const ENRENameKind = ['Norm', 'File', 'Anon', 'Sig', 'Str', 'Num', 'Pvt'] as const;
const ENRENameAnonKind = ['Function', 'Class', 'ArrowFunction', /* Other Language */ 'Namespace', 'Package', 'Struct', 'Union', 'Enum', 'Variable'] as const;
const ENRENameSigKind = ['Callable', 'NumberIndex', 'StringIndex'] as const;

export default class ENREName<T extends typeof ENRENameKind[number]> {
  // Normal name, the payload is an identifier (string).
  // e.g.: foo
  constructor(kind: 'Norm', payload: string);
  // A file name, the payload is its name (base+ext).
  // e.g.: <File foo.js>
  constructor(kind: 'File', payload: string);
  // An anonymous entity, the payload denotes what kind of anonymous entity it is.
  // e.g.: <Anon Function>
  constructor(kind: 'Anon', payload: typeof ENRENameAnonKind[number]);
  // A signature, the payload denotes what kind of signature it is.
  // e.g.: <Sig Callable>
  constructor(kind: 'Sig', payload: typeof ENRENameSigKind[number]);
  // A string, the payload is its content.
  // e.g.: <Str a-not-valid-identifier>
  constructor(kind: 'Str', payload: string);
  // A number, the payload is its raw value (in string), and actual is its evaluated value.
  // e.g.: <Num 1e-10>
  constructor(kind: 'Num', payload: string, actual?: number);
  // A class private member, the payload is its name without prefix `#`.
  // e.g.: <Pvt foo> (Corresponds to `#foo` in source code)
  constructor(kind: 'Pvt', payload: string);
  constructor(
    public kind: T,
    public payload: string,
    public actual?: number,
  ) {
    if (ENRENameKind.indexOf(kind) === -1) {
      throw `Invalid ENREName.kind ${kind}`;
    }

    if (kind === 'Anon' && ENRENameAnonKind.indexOf(payload as any) === -1) {
      throw `Invalid ENREName.payload ${payload} of ENREName.kind Anon`;
    }

    if (kind === 'Sig' && ENRENameSigKind.indexOf(payload as any) === -1) {
      throw `Invalid ENREName.payload ${payload} of ENREName.kind Sig`;
    }

    if (kind === 'Num' && actual === undefined) {
      this.actual = Number(payload);
      if (isNaN(this.actual)) {
        console.warn('ENREName.actual evaluated to NaN for ENREName.kind Num');
      }
    }
  }

  get codeName() {
    if (this.kind === 'Anon') {
      return '';
    } else if (this.kind === 'Sig') {
      return '';
    } else if (this.kind === 'Str') {
      return `'${this.payload}'`;
    } else if (this.kind === 'Num') {
      return `${this.actual}`;
    } else if (this.kind === 'Pvt') {
      return `#${this.payload}`;
    } else {
      return this.payload;
    }
  }

  get codeLength() {
    if (this.kind === 'Anon') {
      return 0;
    } else if (this.kind === 'Sig') {
      return 0;
    } else if (this.kind === 'Str') {
      return this.payload.length + 2;
    } else if (this.kind === 'Num') {
      return this.payload.length;
    } else if (this.kind === 'Pvt') {
      return this.payload.length + 1;
    } else {
      return this.payload.length;
    }
  }

  get string() {
    if (this.kind === 'Norm') {
      return this.payload;
    } else {
      return `<${this.kind} ${this.payload}>`;
    }
  }

  static stringRegex = new RegExp(/<(.+) (.+)>/);

  /**
   * Parse ENRE name formatted string to ENREName object.
   *
   * @param str Formatted string in the shape of `<XXX YYY>` or a normal identifier.
   */
  static fromString(str: string): ENREName<any> {
    const res = this.stringRegex.exec(str);
    if (res === null) {
      return new ENREName('Norm', str);
    } else {
      // @ts-ignore
      return new ENREName(res[1], res[2]);
    }
  }
}
