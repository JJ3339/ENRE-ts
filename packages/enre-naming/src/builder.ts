import {ENRENameFile} from './xml/tag-file';
import {ENRENameAnonymous, ENRENameAnonymousTypes} from './xml/tag-anonymous';
import {ENRENameModified, ENRENameModifiedTypes} from './xml/tag-modified';
import xmlParser from './utils/xml-parser';
import allowedTags from './xml/allowed-tags';
import xmlBuilder from './utils/xml-builder';
import {currentProfile, FormatProfile} from './format-profiles';

/**
 * Rich object representing an identifier name with associated properties, is used by ENRE internally.
 * When running tests, it is the xml-styled `printableName` that to be compared for equality.
 */
export type ENREName = {
  /**
   * Holds all information that any other output view depends on.
   */
  payload: string | ENRENameFile | ENRENameAnonymous | ENRENameModified;
  /**
   * Entity's name that in a normalized form,
   * this will align with the node interpreter.
   */
  codeName: string;
  /**
   * Added with entity's start column will be the range of that entity's the most original cursor selectable range.
   */
  codeLength: number;
  /**
   * Formatted string that print an object as an XML string for disambiguation (by default).
   * This will be used to form entity's qualified name chain.
   */
  printableName: string;
}

/**
 * ENRE uses complex structure to describe code name if it is not simply an identifier.
 *
 * If the code name is an identifier (in most cases), ENREName is just that string;
 * in contrast, an object takes that responsibility.
 *
 * While outputting, string will be printed as-is; object will be printed as an XML string
 * (various examples can be found in the assertion block of documentations).
 * If necessary, ENREName can also be in other formats.
 *
 * This function will also be used in test environment to parse an XML string back to the object.
 */
export const buildENREName = <T extends ENREName['payload'] = string>(payload: T): ENREName => {
  if (typeof payload === 'string') {
    /**
     * Identifiers do not start with '<',
     * so in this case, an XML string is passed in.
     *
     * Since test is written without TypeScript, so there is no need to add `string` type
     * for parameter type if the generic type parameter is not string.
     */
    if (payload.charAt(0) === '<') {
      const res = xmlParser.parse(payload);
      const keys = Object.keys(res);

      if (keys.length === 1 && allowedTags.includes(keys[0] as any)) {
        if (keys[0] === allowedTags[0]) {
          // File
          return buildENREName<ENRENameFile>(res(allowedTags[0]));
        } else if (keys[0] === allowedTags[1]) {
          // Anonymous
          return buildENREName<ENRENameAnonymous>(res[allowedTags[1]]);
        } else if (keys[0] === allowedTags[2]) {
          // Modified
          return buildENREName<ENRENameModified>(res[allowedTags[2]]);
        }
      } else {
        throw new Error(`Invalid ENREName XML string ${payload}`);
      }
    } else {
      return {
        payload,
        codeName: payload,
        codeLength: payload.length,
        printableName: payload,
      };
    }
  } else if (typeof payload === 'object') {
    if ('ext' in payload) {
      const {base, ext} = payload as ENRENameFile;
      const trustedPayload: ENRENameFile = {base, ext};

      return {
        payload: trustedPayload,
        codeName: base + '.' + ext,
        codeLength: base.length + ext.length + 1,
        printableName: xmlBuilder('File', trustedPayload),
      };
    } else if ('as' in payload) {
      if (ENRENameAnonymousTypes.includes(payload.as as any)) {
        const {as} = payload as ENRENameAnonymous;
        const trustedPayload: ENRENameAnonymous = {as};

        return {
          payload: trustedPayload,
          codeName: '',
          codeLength: 0,
          printableName: xmlBuilder('Anonymous', trustedPayload),
        };
      } else if (ENRENameModifiedTypes.includes(payload.as as any)) {
        const {raw, as} = payload as ENRENameModified;
        let trustedPayload: ENRENameModified;
        let codeName: string;
        let codeLength = raw.length;

        if (as === 'NumericLiteral') {
          const {value} = payload as Extract<ENRENameModified, { value: string }>;

          if (!value) {
            throw new Error('Missing property value for NumericLiteral as ENRENameModified');
          }

          trustedPayload = {raw, as, value};
          /**
           * Follows the behavior of node interpreter and JetBrains IDEs,
           * which using normalized number string as printed key or breadcrumb.
           */
          codeName = value;
        } else {
          trustedPayload = {raw, as};

          switch (as) {
            case 'StringLiteral':
              // Add the length of 2 commas.
              codeLength += 2;
              codeName = `'${raw}'`;
              break;
            case 'PrivateIdentifier':
              // Add the length of sharp(#).
              codeLength += 1;
              codeName = `#${raw}`;
              break;
          }
        }

        return {
          payload: trustedPayload,
          codeName,
          codeLength,
          get printableName() {
            switch (currentProfile) {
              case FormatProfile.default:
                return xmlBuilder('Modified', trustedPayload);
              case FormatProfile.understand:
                return raw;
              case FormatProfile.node:
                return codeName;
            }
          },
        };
      } else {
        throw new Error(`Unexpected property value ${payload.as} for ENREName.payload.as`);
      }
    } else {
      throw new Error('Missing required property for ENREName');
    }
  } else {
    throw new Error(`Unsupported type ${typeof payload} as payload of ENREName`);
  }

  throw new Error('Unexpected error encountered while building ENREName');
};
