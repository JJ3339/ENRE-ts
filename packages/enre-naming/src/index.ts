import {XMLParser} from 'fast-xml-parser';
import {currentProfile, FormatProfile} from './format-profiles';

export {usingNewFormatProfile} from './format-profiles';

/**
 * Allowed XML name string tags that will form an XML name string.
 */
const allowedTags = ['Anonymous', 'Modified'] as const;

const xmlParser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: '',
  /**
   * Has to make `allowedTags` mutable for parser to read in, or a TS error will show.
   * The reason to make `allowedTags` `as const` is for the use of exact identifier inference.
   */
  unpairedTags: allowedTags as unknown as Array<string>,
});

const xmlBuilder = (tag: typeof allowedTags[number], props: Record<string, string>) => {
  let tmp = '<' + tag;

  tmp = Object.keys(props).reduce((prev, curr) =>
      prev + ` ${curr}="${props[curr]}"`
    , tmp);

  return tmp + '>';
};

/**
 * Allowed types for `Anonymous` entity name.
 */
export const ENRENameAnonymousTypes = [
  'Function',
  'ArrowFunction',
  'Method',
  'Class',
  'CallableSignature',
  'NumberIndexSignature',
  'StringIndexSignature',

  // In other languages
  'Namespace',
  'Package',
  'Struct',
  'Union',
  'Enum',

  // To support Understand-cpp
  'Variable',
] as const;
export type ENRENameAnonymousType = typeof ENRENameAnonymousTypes[number];

/**
 * Allowed types for `Modified` entity name.
 */
export const ENRENameModifiedTypes = [
  'StringLiteral',
  'NumericLiteral',
  'PrivateIdentifier'
] as const;
export type ENRENameModifiedType = typeof ENRENameModifiedTypes[number];

/**
 * An object representing properties needed by an anonymous entity.
 */
export type ENRENameAnonymous = {
  /**
   * Indicates in which type of anonymous entity this is.
   */
  readonly as: ENRENameAnonymousType;
}

/**
 * An object representing properties needed by a non-identifier entity.
 */
export type ENRENameModified = {
  readonly raw: string;
  readonly as: Exclude<ENRENameModifiedType, 'NumericLiteral'>;
} | {
  readonly raw: string;
  readonly as: 'NumericLiteral';
  /**
   * As for `NumericLiteral`, the value is normalized base 10 number(in string).
   */
  readonly value: string;
}

export const ENRENamePrintableNameFormats = ['xml', 'simple'] as const;
export type ENRENamePrintableNameFormat = typeof ENRENamePrintableNameFormats[number];

/**
 * Rich object representing an identifier name with associated properties, is used by ENRE internally.
 *
 * When running a test, it is the `printableName` that to be compared for equality.
 */
export type ENREName = {
  /**
   * Holds all information any other output view depends on.
   */
  payload: string | ENRENameAnonymous | ENRENameModified;
  /**
   * Entity's name that in a normalized form,
   * this will align with the node interpreter.
   */
  codeName: string;
  /**
   * Added by entity's start location will be the range of that entity's the most original cursor selectable range.
   */
  codeLength: number;
  /**
   * Formatted(in standard of ENRE or Understand depending on the specific configuration) string that
   * print an object as an XML string for disambiguation.
   * This will be used to form entity's qualified name chain.
   */
  printableName: string;
}

/**
 * ENRE uses complex structure to describe code name if it is not simply an identifier.
 *
 * If the code name is an identifier (in most of the case), ENREName will just be that string;
 * whereas if there are some extra information needs to be record, an object takes that responsibility.
 *
 * When print out, string will be printed as is; object will be printed as an XML string
 * (various examples can be found in the meta block of documentation).
 * If necessary, ENREName can also be formatted in the format of Understand, for easier comparison.
 *
 * This function will also be used in tests to parse an XML string back to the object representing the
 * same information.
 */
export const buildENREName = <T extends ENREName['payload'] = string>(payload: T): ENREName => {
  if (typeof payload === 'string') {
    /**
     * Normal identifier will never be started with '<',
     * so in this case, an XML string is passed in.
     *
     * Since test is written without TypeScript, so there is no need to add `string` type
     * for parameter type if the generic type variable is not string.
     */
    if (payload.charAt(0) === '<') {
      const res = xmlParser.parse(payload);
      const keys = Object.keys(res);

      // @ts-ignore
      if (keys.length === 1 && allowedTags.includes(keys[0])) {
        if (keys[0] === allowedTags[0]) {
          // Anonymous
          return buildENREName<ENRENameAnonymous>(res[allowedTags[0]]);
        } else if (keys[0] === allowedTags[1]) {
          // Modified
          return buildENREName<ENRENameModified>(res[allowedTags[1]]);
        }
      } else {
        throw new Error(`Invalid ENREName XML string ${payload}`);
      }
    } else {
      return {
        payload,
        codeName: payload,
        codeLength: payload.length,
        printableName: payload,
      };
    }
  } else if (typeof payload === 'object') {
    // @ts-ignore
    if (ENRENameAnonymousTypes.includes(payload.as)) {
      const {as} = payload as ENRENameAnonymous;
      const trustedPayload: ENRENameAnonymous = {as};

      return {
        payload: trustedPayload,
        codeName: '',
        codeLength: 0,
        printableName: xmlBuilder('Anonymous', trustedPayload),
      };
      // @ts-ignore
    } else if (ENRENameModifiedTypes.includes(payload.as)) {
      const {raw, as} = payload as ENRENameModified;
      let trustedPayload: ENRENameModified;
      let codeName: string;
      let codeLength = raw.length;

      if (as === 'NumericLiteral') {
        const {value} = payload as Extract<ENRENameModified, { value: string }>;

        if (!value) {
          throw new Error('Missing property value for NumericLiteral as ENRENameModified');
        }

        trustedPayload = {raw, as, value};
        /**
         * Follows the behavior of node interpreter and JetBrains IDEs,
         * which using normalized number string as printed key or breadcrumb.
         */
        codeName = value;
      } else {
        trustedPayload = {raw, as};

        switch (as) {
        case 'StringLiteral':
          // Add the length of 2 commas.
          codeLength += 2;
          codeName = `'${raw}'`;
          break;
        case 'PrivateIdentifier':
          // Add the length of sharp(#).
          codeLength += 1;
          codeName = `#${raw}`;
          break;
        }
      }

      return {
        payload: trustedPayload,
        codeName,
        codeLength,
        get printableName() {
          switch (currentProfile) {
          case FormatProfile.default:
            return xmlBuilder('Modified', trustedPayload);
          case FormatProfile.understand:
            return raw;
          case FormatProfile.node:
            return codeName;
          }
        },
      };
    } else {
      throw new Error(`Unexpected property value ${payload.as} for ENREName.payload.as`);
    }
  } else {
    throw new Error(`Unsupported type ${typeof payload} as payload of ENREName`);
  }

  throw new Error('Unexpected error encountered while building ENREName');
};
