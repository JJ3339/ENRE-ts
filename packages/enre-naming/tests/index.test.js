import {buildENREName} from '../src';

describe('Build ENRE name', () => {
  it('with identifier string', () => {
    expect(buildENREName('foo')).toMatchSnapshot();
  });

  describe('with file payload', () => {
    it('js', () => {
      expect(buildENREName({base: 'file0', ext: 'js'})).toMatchSnapshot();
    });

    it('mjs', () => {
      expect(buildENREName({base: 'file0', ext: 'mjs'})).toMatchSnapshot();
    });

    it('cjs', () => {
      expect(buildENREName({base: 'file0', ext: 'cjs'})).toMatchSnapshot();
    });

    it('jsx', () => {
      expect(buildENREName({base: 'file0', ext: 'jsx'})).toMatchSnapshot();
    });

    it('ts', () => {
      expect(buildENREName({base: 'file0', ext: 'ts'})).toMatchSnapshot();
    });

    it('d.ts', () => {
      expect(buildENREName({base: 'file0', ext: 'd.ts'})).toMatchSnapshot();
    });

    it('mts', () => {
      expect(buildENREName({base: 'file0', ext: 'mts'})).toMatchSnapshot();
    });

    it('cts', () => {
      expect(buildENREName({base: 'file0', ext: 'cts'})).toMatchSnapshot();
    });

    it('tsx', () => {
      expect(buildENREName({base: 'file0', ext: 'tsx'})).toMatchSnapshot();
    });

    it('json', () => {
      expect(buildENREName({base: 'file0', ext: 'json'})).toMatchSnapshot();
    });
  });

  describe('with anonymous payload', () => {
    it('as Function', () => {
      expect(buildENREName({as: 'Function'})).toMatchSnapshot();
    });

    it('as ArrowFunction', () => {
      expect(buildENREName({as: 'ArrowFunction'})).toMatchSnapshot();
    });

    it('as Class', () => {
      expect(buildENREName({as: 'Class'})).toMatchSnapshot();
    });
  });

  describe('with computed payload', () => {
    it('as StringLiteral', () => {
      expect(buildENREName({raw: 'foo', as: 'StringLiteral'})).toMatchSnapshot();
    });

    it('as PrivateIdentifier', () => {
      expect(buildENREName({raw: 'foo', as: 'PrivateIdentifier'})).toMatchSnapshot();
    });

    it('as NumericLiteral', () => {
      expect(buildENREName({raw: '123', as: 'NumericLiteral', value: '123'})).toMatchSnapshot();
    });
  });

  describe('with XML string', () => {
    // TODO
  });
});
