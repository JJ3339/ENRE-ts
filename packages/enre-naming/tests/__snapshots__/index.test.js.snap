// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build ENRE name with anonymous payload as ArrowFunction 1`] = `
Object {
  "codeLength": 0,
  "codeName": "",
  "payload": Object {
    "as": "ArrowFunction",
  },
  "printableName": "<Anonymous as=\\"ArrowFunction\\">",
}
`;

exports[`Build ENRE name with anonymous payload as Class 1`] = `
Object {
  "codeLength": 0,
  "codeName": "",
  "payload": Object {
    "as": "Class",
  },
  "printableName": "<Anonymous as=\\"Class\\">",
}
`;

exports[`Build ENRE name with anonymous payload as Function 1`] = `
Object {
  "codeLength": 0,
  "codeName": "",
  "payload": Object {
    "as": "Function",
  },
  "printableName": "<Anonymous as=\\"Function\\">",
}
`;

exports[`Build ENRE name with computed payload as NumericLiteral 1`] = `
Object {
  "codeLength": 3,
  "codeName": "123",
  "payload": Object {
    "as": "NumericLiteral",
    "raw": "123",
    "value": "123",
  },
  "printableName": "<Modified raw=\\"123\\" as=\\"NumericLiteral\\" value=\\"123\\">",
}
`;

exports[`Build ENRE name with computed payload as PrivateIdentifier 1`] = `
Object {
  "codeLength": 4,
  "codeName": "#foo",
  "payload": Object {
    "as": "PrivateIdentifier",
    "raw": "foo",
  },
  "printableName": "<Modified raw=\\"foo\\" as=\\"PrivateIdentifier\\">",
}
`;

exports[`Build ENRE name with computed payload as StringLiteral 1`] = `
Object {
  "codeLength": 5,
  "codeName": "'foo'",
  "payload": Object {
    "as": "StringLiteral",
    "raw": "foo",
  },
  "printableName": "<Modified raw=\\"foo\\" as=\\"StringLiteral\\">",
}
`;

exports[`Build ENRE name with identifier string 1`] = `
Object {
  "codeLength": 3,
  "codeName": "foo",
  "payload": "foo",
  "printableName": "foo",
}
`;
